pkg/storage/mvcc.go

```diff
@@ -4119,25 +4117,56 @@ func MVCCGarbageCollect(
 	iter := rw.NewMVCCIterator(MVCCKeyAndIntentsIterKind, IterOptions{
 		LowerBound: keys[0].Key,
 		UpperBound: keys[len(keys)-1].Key.Next(),
+		KeyTypes:   IterKeyTypePointsAndRanges,
 	})
 	defer iter.Close()
 	supportsPrev := iter.SupportsPrev()
 
+	// To update mvcc stats, we need to go through range tombstones to determine
+	// GCBytesAge. That requires copying current stack of range keys covering
+	// point since returned range key values are unsafe. We try to cache those
+	// copies by checking start of range.
+	var lastRangeTombstoneStart roachpb.Key
+	var rangeTombstoneTss []hlc.Timestamp
+
 	// Iterate through specified GC keys.
 	meta := &enginepb.MVCCMetadata{}
 	for _, gcKey := range keys {
 		encKey := MakeMVCCMetadataKey(gcKey.Key)
-		ok, metaKeySize, metaValSize, _, err := mvccGetMetadata(iter, encKey, meta)
+		ok, metaKeySize, metaValSize, realKeyChanged, err := mvccGetMetadata(iter, encKey, meta)
 		if err != nil {
 			return err
 		}
 		if !ok {
 			continue
 		}
+
+		// If mvccGetMetadata landed on a bare range tombstone for the key it will
+		// synthesize deletion meta. We need to filter this case out to avoid
+		// updating key stats as the key doesn't exist.
+		// Empty realKeyChanged is an indication that there are no values for the
+		// key present, but it may contain an inlined metadata which we filter out
+		// by checking that it is not inlined.
+		// In that case, we can skip to the next gc key as there's nothing to GC.
+		// As a side effect of this change, we should be positioned on the point
+		// key or inlined meta at this point and can do further checks for GC
+		// eligibility.
 		inlinedValue := meta.IsInline()
-		implicitMeta := iter.UnsafeKey().IsValue()
+		if realKeyChanged.IsEmpty() && !inlinedValue {
+			continue
+		}
+
+		// We are guaranteed now to be positioned at the meta or version key that
+		// belongs to gcKey history.
+
+		unsafeKey := iter.UnsafeKey()
+		implicitMeta := unsafeKey.IsValue()
 		// First, check whether all values of the key are being deleted.
 		//
+		// We also catch cases of meta synthesized from range tombstone where the
```

Comment:
I now think that the initial code that just looked on points as they were removed to figure out if all versions are deleted was cleaner and much easier to understand. There are maybe 14 meaningful combinations of how we can put range keys and meta and subsequent keys and what the function returns. Then I just see which combinations of values match which cases, but adding another range above adds a couple more. I'll put a TODO to make it more reasonable.

Reply:
I believe the initial code had the same bug, since it didn't modify this branch (or am I misremembering?).

Reply:
It did, I think usage of metadata is flawed. I think having a metadata object where those if conditions are exposed as boolean methods would be better. Then you can just ask what it is and do action instead of trying to combine values every time you need to make decision. So I moved meta-range tombstone handling to the top so that it doesn't leak to subsequent logic.

Reply:
>I think having a metadata object where those if conditions are exposed as boolean methods would be better.

Yeah, I'd agree with this. I feel like the use of synthetic metadata in general is a bit iffy, but at least centralizing the logic in easy-to-read methods would be a good start. Feel free to attach a commit or submit a follow-up PR.

>So I moved meta-range tombstone handling to the top so that it doesn't leak to subsequent logic.

Good idea, that's clearer.